### ISSUE TITLE
Pyright does not detect a subclass not implementing an abstractmethod if object is constructed via a classmethod

### ISSUE THREAD

**Issue by Kh4l3tH** [microsoft/pyright#10250]
Pyright does not detect, that a subclass is not implementing an abstractmethod if object is instantiated via a classmethod:
```python
from abc import ABC, abstractmethod
from typing import Self


class Base(ABC):
    @abstractmethod
    def method(self): ...


class Sub(Base):
    @classmethod
    def default(cls) -> Self:
        return cls()


Sub.default()   # No errors
# Sub()         # Error: "Base.method" is not implemented
```

The above example produces no errors even though `Sub` is not implementing `method` and fails at runtime.
When an object is instantiated via `Sub()` pyright correctly shows an error.

The issue appears both with pylance (2025.4.1) in VSCode and with pyright (1.1.398) run from command line.

**Comment by erictraut** [comment]
Yes, this is by design. You haven't marked `Sub` as `@final`, so it is possible to subclass it. If you do so, then it's perfectly legal for the class method `default` to call its constructor.

If you want to detect an error like this, you should apply an `@final` decorator to the class.

**Comment by Kh4l3tH** [comment]
> You haven't marked `Sub` as `@final`, so it is possible to subclass it.
True! But i have explicitly called `default` on **`Sub`** which is clearly not implementing `method`!

> If you do so, then it's perfectly legal for the class method `default` to call its constructor.
Yes, but that's not what I did.

> If you want to detect an error like this, you should apply an `@final` decorator to the class.
That works, but isn't pyright supposed to tell me errors about errors before i run the code?
In the code example it's obvious, that it will fail at runtime and it should be obvious for pyright as well, and still you say it is by design that I won't get a notification?!
That sounds very odd to me. What am I missing here?

**Comment by Kh4l3tH** [comment]
@erictraut:

I have updated the snippet so that the use of `@final` is no longer a solution to the problem:

```python
from abc import ABC, abstractmethod
from typing import final, Self

class Base(ABC):
    @abstractmethod
    def method(self): ...

class Mid(Base):
    @classmethod
    def default(cls) -> Self:
        return cls()

@final
class Sub(Mid):
    def method(self):
        pass

Sub.default()   # pyright：ok / python：ok
Mid.default()   # pyright：ok / python：TypeError: Can't instantiate abstract class Mid without an implementation for abstract method 'method'
```

**Comment by erictraut** [comment]
To catch this class of error with a static type checker, you'd need to mark `Mid` as final. If you intend to subclass `Mid` as you've done in the last example, then a static type checker will not report an error even if you attempt to instantiate `Mid`. To catch an error like this, a type checker would need to perform global code analysis rather than local code analysis. Global code analysis is computationally very expensive (infeasible for more complex program), and I'm not aware of any Python type checkers that do this. The Python type system is designed for local type analysis. Each function's implementation is analyzed based on the context of its interface contract. In your example, there are no violations to report from the perspective of local type analysis. Not surprisingly, other type checkers like mypy do not report an error for this code either.

Static type checkers are able to find and report many common classes of bugs in code, but they will not report all potential bugs.
